# Project final:

Team:
- Kiều Hải Đăng - 19127347
- Lã Minh Hiếu - 19127400

## 1. Collecting data:

**1.1 The subject of data:**

- The dataset about TED Talks on the TED.com

**1.2 The origin of data:**
- We get dataset form [kaggle.](https://www.kaggle.com/rounakbanik/ted-talks?select=ted_main.csv)

**1.3 The licence of data:**
- The licence is [CC BY-NC-SA 4.0.](https://creativecommons.org/licenses/by-nc-sa/4.0/)

**1.4 How did authors collect data?**
- The data has been scraped from the official TED Website.

**1.5 Data describtion**
- These datasets contain information of all audio-video recordings of TED Talks uploaded to the official TED.com website until September 21st, 2017. 


## Library

In [23]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import re

## 2. Exploring data

- How many rows and how many columns?
- What is the meaning of each row?

    - Are there **duplicated** rows?
    - What is the meaning of each column?
    - What is the current data type of each column? Are there columns having inappropriate data types?
    
    - With each **numerical** column, how are values distributed?
        - What is the percentage of missing values?
        - Min? max? Are they abnormal?
        
    - With each **categorical** column, how are values distributed?
        - What is the percentage of missing values?
        - How many different values? Show a few
        - Are they abnormal?

### Read data from csv to dataframe 

In [3]:
ted_df = pd.read_csv('ted_main.csv')
ted_df.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

### How many rows and columns does the data have?

In [4]:
num_rows = ted_df.shape[0]
num_cols = ted_df.shape[1]
print('Rows: ',num_rows)
print('Columns: ',num_cols)

Rows:  2550
Columns:  17


### What is the meaning of each row?

Each row represents for a video/audio on ted.com

### Are there duplicated rows

In [9]:
if ted_df[ted_df.duplicated()].shape[0] == 0:
    have_duplicated_rows = False
else:
    have_duplicated_rows = True
print('Are there duplicated rows:', have_duplicated_rows)

Are there duplicated rows: False


### What is the meaning of each column?

There are the description of each column in [kaggle.](https://www.kaggle.com/rounakbanik/ted-talks?select=ted_main.csv)

- **comments**: The number of first level comments made on the talk
- **description**: A blurb of what the talk is about
- **duration**: The duration of the talk in seconds
- **event**: The TED/TEDx event where the talk took place
- **film_date**: The Unix timestamp of filming
- **languages**: The number of languages in which the talk is available
- **main_speaker**: The first named speaker of the talk
- **name**: The official name of the TED Talk. Includes the title and speaker
- **num_speaker**: The number of speakers in the talk
- **published_date**: The Unix timestamp for the publication of the talk on TED.com
- **ratings**: A stringified dictionary of the various ratings given to the talk (inspiring, fascinating, jaw dropping, etc.)
- **related_talk**: A list of dictionaries of recommended talks to watch next
- **speaker_occupation**: The occupation of the main speaker
- **tags**: The themes associated with the talk
- **title**: The title of the talk
- **url**: The URL of the talk
- **views**: The number of views on the talk

### What is the current data type of each column? Are there columns having inappropriate data types?

In [12]:
col_dtypes = ted_df.dtypes
col_dtypes

comments               int64
description           object
duration               int64
event                 object
film_date              int64
languages              int64
main_speaker          object
name                  object
num_speaker            int64
published_date         int64
ratings               object
related_talks         object
speaker_occupation    object
tags                  object
title                 object
url                   object
views                  int64
dtype: object

`ted_df` has 10 columns which have `object` data types: "description","event","main_speaker","name","ratings","related_talks","speaker_occupation","tags","title","title","url". However, "ratings" is a stringified dictionary `{'id': 7, 'name': 'Funny', 'count': 19645}`; "film_date" and "published_date" is `int64` data type.

**We need to preprocessing:**
- "ratings": seperate each element in dictionary to each columns like "Funny","Courageous",etc.
- "film_date" and "published_date": need to turn to datetime

## 3. Preprocessing data

### Transform the data type of "film_date" and "published_date" to dataframe

In [20]:
ted_df['film_date'] = pd.to_datetime(ted_df['film_date'],unit='s')
ted_df['published_date'] = pd.to_datetime(ted_df['published_date'],unit='s')

### "ratings": seperate each element in dictionary to each columns like "Funny","Courageous",etc.

In [45]:
ted_df['ratings'][0]

"[{'id': 7, 'name': 'Funny', 'count': 19645}, {'id': 1, 'name': 'Beautiful', 'count': 4573}, {'id': 9, 'name': 'Ingenious', 'count': 6073}, {'id': 3, 'name': 'Courageous', 'count': 3253}, {'id': 11, 'name': 'Longwinded', 'count': 387}, {'id': 2, 'name': 'Confusing', 'count': 242}, {'id': 8, 'name': 'Informative', 'count': 7346}, {'id': 22, 'name': 'Fascinating', 'count': 10581}, {'id': 21, 'name': 'Unconvincing', 'count': 300}, {'id': 24, 'name': 'Persuasive', 'count': 10704}, {'id': 23, 'name': 'Jaw-dropping', 'count': 4439}, {'id': 25, 'name': 'OK', 'count': 1174}, {'id': 26, 'name': 'Obnoxious', 'count': 209}, {'id': 10, 'name': 'Inspiring', 'count': 24924}]"

## 3. Asking meaningful questions

- What is the question?
- What are benefits of finding the answer?

## 4. Preprocessing + analyzing data to answer each question

## 5. Conclusion

## 6. Reflection

- Each member: What difficulties have you encountered?
- Each member: What have you learned?
- Your group: If you had more time, what would you do?

## 7. Reference